### PR TITLE
Feature/vector inference

### PR DESCRIPTION
**Pull request by blondered**
- Optimized inference for vector models with DOT and COSINE distance using implicit library inference methods

### CODE REVIEW

File: rectools/models/vector.py [hunk 1]
```diff
@@ -43,6 +46,121 @@ class Factors:
...
+
+    def _process_implicit_scores(
+        self, subject_ids: np.ndarray, ids: np.ndarray, scores: np.ndarray
+    ) -> tp.Tuple[InternalIds, InternalIds, Scores]:
+
+        all_target_ids = []
+        all_reco_ids: tp.List[np.ndarray] = []
+        all_scores: tp.List[np.ndarray] = []
+
+        for i in range(scores.shape[0]):
```
Comment (feldlime) [inline rectools/models/vector.py:93]: image
(Naming is not suggested)

File: tests/models/test_vector.py [hunk 2]
```diff
@@ -151,14 +151,14 @@ def test_with_biases(
...
-        assert list(scores) == sum(expected_scores, [])
+        np.testing.assert_almost_equal(scores, np.array(expected_scores).ravel(), decimal=5)
 
     @pytest.mark.parametrize("method", ("u2i", "i2i"))
     def test_with_incorrect_distance(self, method: str) -> None:
         with pytest.raises(ValueError):
             if method == "u2i":
                 m = self.make_model(self.user_biased_factors, self.item_biased_factors, u2i_distance=7)  # type: ignore
-                m._get_u2i_calculator(self.stub_dataset)
+                _, __ = m._get_u2i_vectors(self.stub_dataset)
```
Comment (feldlime) [inline tests/models/test_vector.py:161]: why do you need `_, __ = `?

File: rectools/models/vector.py [hunk 3]
```diff
@@ -43,6 +46,121 @@ class Factors:
...
+            raise ValueError(f"ImplicitRanker is not suitable for {distance} distance")
+        self.distance = distance
+        self.subjects_factors = subjects_factors.astype(np.float32)
+        self.objects_factors = objects_factors.astype(np.float32)
+
+        self.subjects_norms: np.ndarray
+        self.objects_norms: np.ndarray
+        if distance == Distance.COSINE:
+            self.subjects_norms = np.linalg.norm(subjects_factors, axis=1).astype(np.float32)
+            self.objects_norms = np.linalg.norm(objects_factors, axis=1).astype(np.float32)
```
[the remark this answers is not in the document]
Comment (blondered) [reply rectools/models/vector.py]: fixed

File: rectools/models/vector.py [hunk 4]
```diff
@@ -43,6 +46,121 @@ class Factors:
...
+
+    def _get_neginf_score(self) -> float:
```
[the remark this answers is not in the document]
Comment (blondered) [reply rectools/models/vector.py:76]: fixed. i did the test that -np.finfo(batch_distances.dtype).max is equal to current implicit behaviour. not much we can do if smth changes in imlplicit logic in the future

File: rectools/models/vector.py [hunk 5]
```diff
@@ -43,6 +46,121 @@ class Factors:
...
+        dummy_factors = np.array([[1, 2]], dtype=np.float32)
+        neginf = implicit.cpu.topk.topk(  # pylint: disable=c-extension-no-member
+            items=dummy_factors,  # overall whitelist item factors
+            query=dummy_factors,  # query_factors
+            k=1,
+            filter_items=np.array([0]),
+        )[1][0][0]
+        return neginf
+
+    def _get_mask_for_correct_scores(self, scores: np.ndarray) -> tp.List[bool]:
```
[the remark this answers is not in the document]
Comment (blondered) [reply rectools/models/vector.py:79]: `done

File: rectools/models/vector.py [hunk 6]
```diff
@@ -43,6 +46,121 @@ class Factors:
...
+        num_masked = 0
+        min_score = self._get_neginf_score()
+        for el in np.flip(scores):
+            if el <= min_score:
+                num_masked += 1
+            else:
+                break
+        return [True for _ in range(len(scores) - num_masked)] + [False for _ in range(num_masked)]
```
[the remark this answers is not in the document]
Comment (blondered) [reply rectools/models/vector.py]: fixed

File: rectools/models/vector.py (hunk 1 above)
Comment (blondered) [reply rectools/models/vector.py]: fixed

File: tests/models/test_vector.py (hunk 2 above)
Comment (blondered) [reply tests/models/test_vector.py]: fixed

Comment (feldlime) [reply tests/models/test_vector.py]: still, why do you need `_`?

Comment (blondered) [reply tests/models/test_vector.py]: fixed

File: rectools/models/vector.py [hunk 7]
```diff
@@ -43,6 +46,128 @@ class Factors:
...
+
+        subject_factors = self.subjects_factors[subject_ids]
+
+        object_norms = None  # for DOT distance
+        if self.distance == Distance.COSINE:
+            object_norms = self.objects_norms
+            if sorted_item_ids_to_recommend is not None:
+                object_norms = object_norms[sorted_item_ids_to_recommend]
+
+        if object_norms is not None:  # prevent zero division
```
Comment (feldlime) [inline rectools/models/vector.py:147]: you can do this under prev. `if`

Comment (blondered) [reply rectools/models/vector.py]: moved to __init__

File: rectools/models/vector.py [hunk 8]
```diff
@@ -72,6 +72,8 @@ def __init__(self, distance: Distance, subjects_factors: np.ndarray, objects_fac
         if distance == Distance.COSINE:
             self.subjects_norms = np.linalg.norm(self.subjects_factors, axis=1)
             self.objects_norms = np.linalg.norm(self.objects_factors, axis=1)
+            self.objects_norms[self.objects_norms == 0] = 1e-10
```
Comment (feldlime) [inline rectools/models/vector.py:75]: Let's please move it to constant like `EPS = 1e-10`

**Review by feldlime — APPROVED** [review]

**Pull request merged**